### PR TITLE
sundials: assert that lapack and blas are compatible

### PR DESCRIPTION
**Pull request by doronbehar**
Instead of asserting that lapack and blas are both not using 64 bit index size, only make sure they are index size compatible. Plus, do it when evaluating buildInputs, per: https://github.com/NixOS/nixpkgs/pull/94892#discussion_r471110250 

Don't pass SUNDIALS_INDEX_TYPE=int32_t if blas && lapack are compiled
with 64 bit index size.

Bonus: Use checkTarget and not an overrided checkPhase, per:
https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/explicit-phases.md

###### Motivation for this change

Implement stuff I learned at https://savannah.gnu.org/bugs/?57591#comment112

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by GuillaumeDesforges** [comment]
LGTM

Please note that I am no specialist of sundials (and no longer use it (have I ever? oh right it was a dependency for some Haskell lib)).
IMO @idontgetoutmuch is still the maintainer and should approve this.

**Comment by idontgetoutmuch** [comment]
LGTM

**Review by idontgetoutmuch — APPROVED** [review]

**Comment by doronbehar** [comment]
Thanks.

**Pull request merged**
